pkg/controller/pkg/revision/establisher.go

```diff
@@ -50,6 +50,9 @@ type Establisher interface {
 	// GetResourceRefs returns references to all objects that have had ownership
 	// or control established.
 	GetResourceRefs() []runtimev1alpha1.TypedReference
+
+	// Reset clears all cached objects and references.
+	Reset()
 }
```

Comment:
It seems like this interface consists of four methods that must in practice be called in order.  Could this interface (and its API implementation) be merged into one method? Said method might not meet our cyclomatic complexity goals, but I suspect that would be a reasonable tradeoff because it would make it simpler to call and prevent us from having to keep state (i.e. allObjs and resourceRefs).

I'm imagining calling this interface would work something like:

```go
type Establisher interface {
	// The 'control' argument can be inferred from 'pr'.
	Establish(ctx context.Context, o []runtime.Object, pr *v1alpha1.PackageRevision) ([]runtimev1alpha1.TypedReference, error)
}

// Nit - I prefer r.objects.Establish because it reads like "establish objects".
refs, err := r.objects.Establish(ctx, pkg.GetObjects(), pr)
if resource.IsNotControllable(err) {
	log.Debug(errEstablishControl, "error", err)
	r.record.Event(pr, event.Warning(errEstablishControl, err))
	pr.SetConditions(v1alpha1.Unhealthy())
	return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
}
if err != nil {
	log.Debug(errEstablishControl, "error", err)
	r.record.Event(pr, event.Warning(errEstablishControl, err))
	pr.SetConditions(v1alpha1.Unhealthy())
	return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
}
```

Reply:
This is a very reasonable request. For completeness, I would like to explore two other options for your feedback:
1. The first one would be the obvious leave as is. I think my reasoning for originally creating this interface with multiple methods was to potentially allow for custom things to happen between checking for and actually establishing control. We don't have a great use case for it right now (which may be justification in itself to simplify it), but I don't know if it could come into play later when we do some more sophisticated things with dependencies. The other reason for separating it out is that I don't think it is entirely unrealistic that this interface could be used elsewhere, as the pattern of making sure you can control _all_ before actually modifying _any_ seems like a pretty generic use case.
2. My immediate, and potentially misguided, reaction with a single method interface that is not _currently_ being used elsewhere is just to put the logic directly in the caller (e.g. in the reconcile loop here). I do like the readability of just saying `establish` and moving on from a code readability standpoint, but I'm not sure if it is obscuring something that should actually be plain to anyone reading the reconcile loop.

Anyway, I am pretty comfortable with your suggestion here, just wanted to float these other outlooks by you in case they were something you hadn't considered :+1:

Reply:
I feel that breaking this out into a pluggable interface with one method is the best compromise here. It allows us to test the establishment logic as a smaller, distinct unit with clear inputs and outputs rather than having to have one `Reconcile` method with a high level of complexity. I'm hopeful that breaking this out into a single method won't hurt the readability of the `Reconcile`, but I know this can be a bit subjective and that some people prefer to read all the logic in one place.

Reply:
Seems reasonable to me :+1: